In [1]:
from sklearn.datasets import load_diabetes
from sklearn import tree
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

# Изучение датасета

Будем использовать стандартный набор данных, содержащий информацию о пациентах с диабетом, который предоставляется в библиотеке `Sklearn`.

In [2]:
print(load_diabetes().DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

Датасет состоит из `442` строк и `11` столбцов:
- `age` (возраст пациента);
- `sex` (пол пациента);
- `bmi` (индекс массы тела);
- `bp` (среднее артериальное давление);
- `s1` - `s6` (шесть измерений сыворотки крови);
- `target` (целевой признак, количественная мера прогрессирования заболевания через год после исходного уровня).

In [3]:
df = load_diabetes(as_frame=True).frame

df.sample(5)

age       sex       bmi        bp        s1        s2        s3  \
207  0.009016 -0.044642  0.045529  0.028758  0.012191 -0.013840  0.026550   
334 -0.060003  0.050680 -0.047163 -0.022885 -0.071743 -0.057681 -0.006584   
317  0.019913 -0.044642 -0.034229  0.055165  0.067229  0.074155 -0.006584   
9   -0.070900 -0.044642  0.039062 -0.033213 -0.012577 -0.034508 -0.024993   
412  0.074401 -0.044642  0.085408  0.063187  0.014942  0.013091  0.015505   

           s4        s5        s6  target  
207 -0.039493  0.046133  0.036201   202.0  
334 -0.039493 -0.062917 -0.054925    72.0  
317  0.032833  0.024730  0.069338   190.0  
9   -0.002592  0.067737 -0.013504   310.0  
412 -0.002592  0.006207  0.085907   261.0

Столбцы содержат значения, которые могут показаться необычными. Например, в столбцах `age` и `sex` используются вещественные числа вместо ожидаемых целочисленных значений или категорий. Вероятно, такое преобразование и выбор необычных значений были сделаны с целью анонимизации данных и защиты конфиденциальности пациентов.

# Подготовка данных для обучения модели

Отделим целевой признак от датасета.

In [4]:
features = df.drop('target', axis=1)

target = df.target

print(features.shape)
print(target.shape)

(442, 10)
(442,)


Разделим датасет на тренировочную и валидационную части.

In [5]:
features_train, features_val, target_train, target_val = train_test_split(features, target, random_state=5)

In [6]:
print(f'Размер тренировочной выборки: {features_train.shape}')
print(f'Размер валидационной выборки: {features_val.shape}')

Размер тренировочной выборки: (331, 10)
Размер валидационной выборки: (111, 10)


# Обучение модели дерева решений и оценка ее качества

Будем решать задачу регрессии, так как целевая переменная `target` представляет собой числовую величину.

Попробуем применить дерево решений для решения задачи регрессии.

In [7]:
model = tree.DecisionTreeRegressor()

model.fit(features_train, target_train)

predictions_val = model.predict(features_val)

Для оценки адекватности выбранной модели будем использовать метрику `R2` (коэффициент детерминации), которая сравнивает модель с простой моделью, где все целевые значения в обучающей выборке являются средними.

In [8]:
print(f'Коэффициент детерминации: {r2_score(target_val, predictions_val)}')

Коэффициент детерминации: -0.13038615149756927


Коэффициент детерминации имеет отрицательное значение, что свидетельствует о крайне низком качестве модели. Фактически, такая модель показывает худшие результаты, чем простая модель.

# Обучение модели линейной регрессии и оценка ее качества

Попробуем линейную регрессию.

In [9]:
model = linear_model.LinearRegression()

model.fit(features_train, target_train)

predictions_val = model.predict(features_val)

Оценим адекватность модели.

In [10]:
print(f'Коэффициент детерминации: {r2_score(target_val, predictions_val)}')

Коэффициент детерминации: 0.5308133103536823


Линейная регрессия демонстрирует достойные результаты. Положительное значение метрики `R2` свидетельствует о том, что модель превосходит простую модель, которая была обучена на средних значениях.

# Вывод

В ходе работы была изучена документация `scikit-learn` и проведена подготовка данных для обучения двух моделей машинного обучения — дерева решений и линейной регрессии. После этого был проведен сравнительный анализ результатов обеих моделей. Результаты этого анализа показали, что линейная регрессия превзошла дерево решений и продемонстрировала лучшие результаты.